Add Swagger codegen in angular

jmandel opened the pull request:
This is a PR on top of yours @danqrodney, taking a quick look at codegen in angular2.

There's no obvious/clear framework-sanctioned way to hook a codegen step into the build process, so here I add an explicit `npm run codegen` script, and call it automatically when doing `npm run build`, which should help prevent things from getting too far out of sync. 

For examples of the generated files:
https://gist.github.com/jmandel/c951f5d685d097c7d39e8cc5304a6a78

The files are reasonable, including magic like:

        constructor(protected http: Http, @Optional()@Inject(BASE_PATH) basePath: string, @Optional() configuration: Configuration) {

which allows you to inject different base URLs into the service depending on which environment you're using. I've tried this technique with success (but not included in this PR, because I don't want to clutter up with experimental code).

markfickett commented:
Cool!

What do you think about this v. the auth approach I was setting up? I started by providing an `AuthenticatedHttp` separate from plain old `Http`. Maybe providing `AuthenticatedHttp` in place of normal `Http` is the right thing to do, and then these Swagger-generated services can work without modification. Then nobody can ask for the default `Http` but that's probably OK.

jmandel commented:
Just updated the PR to use the latest version of angular2 codegen (it's under active development) and updated our swagger file to indicate OAuth security enabled on the cohorts endpoint. This yields generated code like https://gist.github.com/jmandel/65b24d30039978f24c388b8a972479d2 which includes the ability to [inject an access token via config](https://gist.github.com/jmandel/65b24d30039978f24c388b8a972479d2#file-cohortsapi-ts-L115-L120). So I think if we set thinkgs up properly, we might be able use this without needing a special wrapper type around `Http`.

danqrodney commented on ui/package.json:
```diff
@@ -5,10 +5,11 @@
     "lint": "ng lint",
-    "e2e": "ng e2e"
+    "e2e": "ng e2e",
+    "codegen": "java -jar ../tools/swagger-codegen-cli.jar generate --lang typescript-angular  --input-spec ../api/src/main/resources/workbench.yaml  --output src/generated --additional-properties ngVersion=2"
```
Does there need to be a "clean" command which removes old generated files?

Or does running this again clear out all old generated files?

markfickett replied:
Yes, we'll add a "clean" step.

markfickett commented:
This is subsumed by a new PR I'm working on with Josh.
